journals/week4.md

```diff
@@ -0,0 +1,320 @@
+flux.content-cache.%d.valid:%d|g\n\
+flux.content-cache.%d.dirty:%d|g\n\
+flux.content-cache.%d.size:%d|g\n\
+flux.content-cache.%d.flush-batch-count:%d|g",
+		rank, (int) zhashx_size (cache->entries),
+		rank, cache->acct_valid,
+		rank, cache->acct_dirty,
+		rank, cache->acct_size,
+		rank, cache->flush_batch_count);
+}
+```
+
+There's not much different here from the message counters, but it is worth noting that all of the stats are being sent as gauges `g` and not counters `C`. I did some testing using both types for each stat, and they all produced similar results, but I think it makes more sense to have the m sent as snapshots of what the value is at each iteration as opposed to the change since the previous iteration.
+
+**Dirty Entries and Flush Batch Count**
+
+![content cache dirty and batch count](../img/gr_flux_cc_bcd_gc.png)
+
+**Count and Valid Entries**
+
+![content cache count and valid](../img/gr_flux_cc_cv_gc.png)
+
+**Cache Size**
+
+![content cache size](../img/gr_flux_cc_s_gc.png)
+
+After I got the code to send the stats setup within the content cache, I tested it by running 25,000 jobs using the `src/throughput.py`.
+
+**Cache Stats**
+
+![content cache stats](../img/gr_flux_cc_a.png)
+
+As you can tell, the size of the cache is much much larger than all of the rest of the stats as it's multiple megabytes being sent in bytes, so let's take a look at the other stats individually to see how we want to scale it.
+
+**Count and Valid Entries**
+
+![content cache count and valid](../img/gr_flux_cc_cv.png)
+
+The count and number of valid entries in the cache are in the hundreds of thousands, so let's start by scaling down the cache size by 100x (1/100).
+
+**Count and Valid Entries and Size * (1/100)**
+
+![content cache count, size, and valid](../img/gr_flux_cc_cvs.png)
+
+
+**Dirty Entries and Flush Batch Count**
+
+![content cache dirty and batch count](../img/gr_flux_cc_bcd.png)
+
+The flush batch count and number of dirty entries are both in the 10's, so let scale the count and valid entries down by 1,000x and the size by 100,000x.
+
+**Stats Scaled down**
+
+![content cache stats scaled down](../img/gr_flux_cc_a_s.png)
+
+Looking back it would probably make more sense to scale the size down by 1024 ^ 2 to get it into megabytes from bytes.
+
+---
+
+#### Timing Asynchronous Calls
+
+Since load and store requests to the cache can be asynchronous, I wanted to track the time it took for each of the calls to complete. Within the content cache, I started a timer at the beginning of each load and store request, stored it in the flux aux storage, and clocked and sent the delta time once each load/store request was completed.
+
+```c
+static  int  cache_store (struct  content_cache  *cache, struct  cache_entry  *e)
+{
+	struct  timespec  *t0;
+	if (!(t0  =  malloc (sizeof (*t0)))) {
+		flux_log (cache->h, LOG_ALERT, "fripp time start");
+	} else {
+		monotime (t0);
+		flux_aux_set (cache->h, "fripp::cc-store", t0, free);
```

Comment:
I think you meant to store this in the message?   There is a `flux_msg_aux_set()` and `_get()`.